onnxruntime/core/providers/qnn/builder/opbuilder/argmax_min_op_builder.cc

```diff
@@ -33,25 +34,26 @@ class ArgMaxMinOpBuilder : public BaseOpBuilder {
                                      const NodeUnit& node_unit,
                                      std::vector<std::string>&& input_names,
                                      const logging::Logger& logger,
-                                     bool is_quantized_model,
+                                     bool is_quantized_node,
                                      bool do_op_validation) const override ORT_MUST_USE_RESULT;
 };
 
 Status ArgMaxMinOpBuilder::IsOpSupported(QnnModelWrapper& qnn_model_wrapper,
                                          const NodeUnit& node_unit,
                                          const logging::Logger& logger,
-                                         bool is_quantized_model) const {
+                                         bool is_npu_backend,
+                                         bool is_quantized_node) const {
   // ONNX ArgMax/ArgMin ops output int64 indices, but the equivalent QNN ops output uint32 indices.
   // The QNN HTP backend does not generally support the int64 type, but QNN EP can just use the uint32 type
   // for ArgMax/ArgMin ops within the graph. However, if the ArgMin/ArgMax op **generates** a graph output,
   // then we cannot support it on the HTP backend.
-  if (is_quantized_model) {
+  if (is_npu_backend) {
```

Comment:
should we start considering the possibility of multiple backends that may have different constraints?
e.g. we already know there exists an htp16 backend in addition to htp backend. maybe better to be specific rather than generically say npu.

Reply:
along this line of thought, is there a way we can make the api future proof such that we don't need to change the signatures again (and touch many files) in the future? 
maybe use a struct or some api to abstract it?

Reply:
Some thoughts: 
- Instead of `is_npu_backend`, we pass in an enum parameter that denotes the specific "target" backend (e.g., `QnnBackend::kHTP`, `QnnBackend::kCPU`). 
- Remove the `is_quantized_node` parameter entirely. Everywhere we're passing in `is_quantized_node`, we are also passing in either a `NodeUnit` or a `NodeUnitIODef`, which the builders can inspect to determine if the node is in QDQ form.

The builders can use the node unit information and the target backend to validate and build the ops. For example, the Slice operator would allow both QDQ and non-QDQ Slice operators for the HTP backend, but only non-QDQ for the cpu backend.

This refactoring would require more care when dealing with the base_op_builder base methods (e.g., ProcessOutputs) since these base methods are operator-type agnostic, but it looks doable (perhaps in a separate PR?).

Reply:
is_quantized_node -- right, I did consider that. Then chose to compare it once and pass to Op builder instead of do it for every Op builder. But like you said, it can make the OpBuilder interface clean if we do it in every op builder. also the cost to do that is much. and considering that customer would enable context binary cache for final deployment. I agree we should remove it.
is_npu_backend -- I was also considering remove it also. It was used for 2 Op builder to exclude non-float32 type support for CPU. That should be covered by Qnn Opvalidation. But I'm not clear how it would impact HTP FP16 yet.

---
In reply to: [1300457143](https://github.com/microsoft/onnxruntime/pull/17194#discussion_r1300457143) [](ancestors = 1300457143)